core/sbt-cloudflow/src/main/scala/cloudflow/sbt/CloudflowFlinkPlugin.scala

```diff
@@ -0,0 +1,240 @@
+        from(CloudflowFlinkDockerBaseImage)
+        user(userInImage)
+
+        copy(depJarsDir, optAppDir, chown = userAsOwner(userInImage))
+        copy(appJarsDir, optAppDir, chown = userAsOwner(userInImage))
+        runRaw(s"cp ${optAppDir}cloudflow-runner.jar  /opt/flink/flink-web-upload/cloudflow-runner.jar")
+        label(streamletDescriptorsLabelName, streamletDescriptorsLabelValue)
+      }
+    },
+
+    build := showResultOfBuild.dependsOn(
+      docker.dependsOn(
+        checkUncommittedChanges
+      )
+    ).value,
+
+    buildAndPublish := showResultOfBuildAndPublish.dependsOn(
+      dockerBuildAndPush.dependsOn(
+        checkUncommittedChanges,
+        verifyDockerRegistry
+      )
+    ).value,
+
+    fork in Compile := true
+  )
+
+  private val verifyDockerRegistry = Def.task {
+    cloudflowDockerRegistry.value.getOrElse(throw DockerRegistryNotSet)
+  }
+
+  private val checkUncommittedChanges = Def.task {
+    val log = streams.value.log
+    if (cloudflowBuildNumber.value.hasUncommittedChanges) {
+      log.warn(s"You have uncommitted changes in ${thisProjectRef.value.project}. Please commit all changes before publishing to guarantee a repeatable and traceable build.")
+    }
+  }
+
+  private val showResultOfBuild = Def.task {
+    val log = streams.value.log
+    val imagePushed = (imageNames in docker).value.head // assuming we only build a single image!
+
+    log.info(" ") // if you remove the space, the empty line will be auto-removed by SBT somehow...
+    log.info("Successfully built the following Cloudflow application image:")
+    log.info(" ")
+    log.info(s"  $imagePushed")
+    log.info(" ")
+    log.info("Before you can deploy this image to a Kubernetes cluster you will")
+    log.info("need to push it to a docker registry that is reachable from the cluster.")
+    log.info(" ")
+  }
+
+  private val showResultOfBuildAndPublish = Def.task {
+    val log = streams.value.log
+    val imagePushed = (imageNames in docker).value.head // assuming we only build a single image!
+
+    log.info(" ") // if you remove the space, the empty line will be auto-removed by SBT somehow...
+    log.info("Successfully built and published the following Cloudflow application image:")
+    log.info(" ")
+    log.info(s"  $imagePushed")
+    log.info(" ")
+    log.info("You can deploy the application to a Kubernetes cluster using any of the the following commands:")
+    log.info(" ")
+    log.info(s"  kubectl cloudflow deploy $imagePushed")
+    log.info(" ")
+  }
+
+  def zlibCompression(raw: Array[Byte]): Array[Byte] = {
+    val deflater = new Deflater()
+    val compressed = new ByteArrayOutputStream(0)
+    deflater.setInput(raw)
+    deflater.finish()
+    val buffer = new Array[Byte](1024)
+    while (!deflater.finished()) {
+      val len = deflater.deflate(buffer)
+      compressed.write(buffer, 0, len)
+    }
+    deflater.end()
+    compressed.toByteArray()
+  }
+
+  def buildStreamletDescriptors(
```

Comment:
This can be removed.

Reply:
Why ? We still need to build the descriptors - right ? Removing the imageName though ..

Reply:
This method only adds the imageName stuff, it does nothing else I think?

Reply:
The image name stuff is gone .. it now does this ..

```scala
def buildStreamletDescriptors(
  detectedStreamlets: Map[String, Config]): Def.Initialize[Task[Iterable[StreamletDescriptor]]] = Def.task {
    val detectedStreamletDescriptors = detectedStreamlets.map {
    case (_, configDescriptor) ⇒
      val jsonString = configDescriptor.root().render(ConfigRenderOptions.concise())
      jsonString.parseJson.convertTo[cloudflow.blueprint.StreamletDescriptor]
  }
  detectedStreamletDescriptors
}
```
generates the descriptors from the Map ..

Reply:
ah ok, looks good.